#### Import and explore the data


In [11]:
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('max_rows', None)


import numpy as np
import seaborn as sns
import statsmodels.api as sm
import matplotlib.pyplot as plt
#%matplotlib inline
sns.set(style="whitegrid")


import warnings 
warnings.filterwarnings('ignore')

##### Reading Data

In [17]:
#path = "/content/drive/MyDrive/Colab Notebooks/Greatelearning/"
path = "/Data/"
data = pd.read_csv(path+"signal-data.csv")
data.head()

Time        0        1          2          3       4      5  \
0  2008-07-19 11:55:00  3030.93  2564.00  2187.7333  1411.1265  1.3602  100.0   
1  2008-07-19 12:32:00  3095.78  2465.14  2230.4222  1463.6606  0.8294  100.0   
2  2008-07-19 13:17:00  2932.61  2559.94  2186.4111  1698.0172  1.5102  100.0   
3  2008-07-19 14:43:00  2988.72  2479.90  2199.0333   909.7926  1.3204  100.0   
4  2008-07-19 15:22:00  3032.24  2502.87  2233.3667  1326.5200  1.5334  100.0   

          6       7       8       9      10      11        12   13       14  \
0   97.6133  0.1242  1.5005  0.0162 -0.0034  0.9455  202.4396  0.0   7.9558   
1  102.3433  0.1247  1.4966 -0.0005 -0.0148  0.9627  200.5470  0.0  10.1548   
2   95.4878  0.1241  1.4436  0.0041  0.0013  0.9615  202.0179  0.0   9.5157   
3  104.2367  0.1217  1.4882 -0.0124 -0.0033  0.9629  201.8482  0.0   9.6052   
4  100.3967  0.1235  1.5031 -0.0031 -0.0072  0.9569  201.9424  0.0  10.5661   

         15       16      17        18       19      20       21       22  \
0  414.8710  10.0433  0.9680  192.3963  12.5190  1.4026 -5419.00  2916.50   
1  414.7347   9.2599  0.9701  191.2872  12.4608  1.3825 -5441.50  2604.25   
2  416.7075   9.3144  0.9674  192.7035  12.5404  1.4123 -5447.75  2701.75   
3  422.2894   9.6924  0.9687  192.1557  12.4782  1.4011 -5468.25  2648.25   
4  420.5925  10.3387  0.9735  191.6037  12.4735  1.3888 -5476.25  2635.25   

        23       24      25      26      27       28      29      30      31  \
0 -4043.75   751.00  0.8955  1.7730  3.0490  64.2333  2.0222  0.1632  3.5191   
1 -3498.75 -1640.25  1.2973  2.0143  7.3900  68.4222  2.2667  0.2102  3.4171   
2 -4047.00 -1916.50  1.3122  2.0295  7.5788  67.1333  2.3333  0.1734  3.5986   
3 -4515.00 -1657.25  1.3137  2.0038  7.3145  62.9333  2.6444  0.2071  3.3813   
4 -3987.50   117.00  1.2887  1.9912  7.2748  62.8333  3.1556  0.2696  3.2728   

        32      33       34       35       36       37       38        39  \
0  83.3971  9.5126  50.6170  64.2588  49.3830  66.3141  86.9555  117.5132   
1  84.9052  9.7997  50.6596  64.2828  49.3404  64.9193  87.5241  118.1188   
2  84.7569  8.6590  50.1530  64.1114  49.8470  65.8389  84.7327  118.6128   
3  84.9105  8.6789  50.5100  64.1125  49.4900  65.1951  86.6867  117.0442   
4  86.3269  8.7677  50.2480  64.1511  49.7520  66.1542  86.1468  121.4364   

      40     41    42        43       44        45        46      47  \
0  61.29  4.515  70.0  352.7173  10.1841  130.3691  723.3092  1.3072   
1  78.25  2.773  70.0  352.2445  10.0373  133.1727  724.8264  1.2887   
2  14.37  5.434  70.0  364.3782   9.8783  131.8027  734.7924  1.2992   
3  76.90  1.279  70.0  363.0273   9.9305  131.8027  733.8778  1.3027   
4  76.39  2.209  70.0  353.3400  10.4091  176.3136  789.7523  1.0341   

         48   49        50        51   52     53     54      55      56  \
0  141.2282  1.0  624.3145  218.3174  0.0  4.592  4.841  2834.0  0.9317   
1  145.8445  1.0  631.2618  205.1695  0.0  4.590  4.842  2853.0  0.9324   
2  141.0845  1.0  637.2655  185.7574  0.0  4.486  4.748  2936.0  0.9139   
3  142.5427  1.0  637.3727  189.9079  0.0  4.486  4.748  2936.0  0.9139   
4  138.0882  1.0  667.7418  233.5491  0.0  4.624  4.894  2865.0  0.9298   

       57      58       59        60       61        62       63       64  \
0  0.9484  4.7057  -1.7264  350.9264  10.6231  108.6427  16.1445  21.7264   
1  0.9479  4.6820   0.8073  352.0073  10.3092  113.9800  10.9036  19.1927   
2  0.9447  4.5873  23.8245  364.5364  10.1685  115.6273  11.3019  16.1755   
3  0.9447  4.5873  24.3791  361.4582  10.2112  116.1818  13.5597  15.6209   
4  0.9449  4.6414 -12.2945  355.0809   9.7948  144.0191  21.9782  32.2945   

        65        66      67        68   69        70       71        72  \
0  29.5367  693.7724  0.9226  148.6009  1.0  608.1700  84.0793       NaN   
1  27.6301  697.1964  1.1598  154.3709  1.0  620.3582  82.3494       NaN   
2  24.2829  710.5095  0.8694  145.8000  1.0  625.9636  84.7681  140.6972  

##### Exploratory Data Analysis



###### shape of the data

In [18]:
data.shape

(1567, 592)

Observation
- there are 592 features and 1500+ observations

###### rename column name

In [19]:
data_analysis = data.copy()
data_analysis = data_analysis.add_prefix("dimension_")
data_analysis.rename(columns={"dimension_Pass/Fail": "Target"},inplace = True)
data_analysis.columns

Index(['dimension_Time', 'dimension_0', 'dimension_1', 'dimension_2',
       'dimension_3', 'dimension_4', 'dimension_5', 'dimension_6',
       'dimension_7', 'dimension_8',
       ...
       'dimension_581', 'dimension_582', 'dimension_583', 'dimension_584',
       'dimension_585', 'dimension_586', 'dimension_587', 'dimension_588',
       'dimension_589', 'Target'],
      dtype='object', length=592)

###### Treating Target column

In [20]:
print(data_analysis["Target"].isna().sum()) # No missing value in target 
print(data_analysis["Target"].value_counts())

# Update Fail (1) = 0 and Pass (-1) = 1 for the better understanding
data_analysis["Target"] = np.where(data_analysis["Target"] == -1,1,0 )
print(data_analysis["Target"].value_counts())

print(round(data_analysis['Target'].value_counts() / data_analysis['Target'].count(),2)*100)

0
-1    1463
 1     104
Name: Target, dtype: int64
1    1463
0     104
Name: Target, dtype: int64
1    93.0
0     7.0
Name: Target, dtype: float64


 - Observation 
  - To avoid cognitive load , target value -1 is converted to 1 ( which represent pass) and 1 is converted to 0 ( which is fail)
  - 93% of observation represent the pass and 7 % represent fail

###### Missing data Stat

In [30]:
print("# of data missing in Obsewrvation",data_analysis.isna().sum().sum())
print("# of data Features has missing Value",data_analysis.isna().sum()[data_analysis.isna().sum() >=1].count())

# of data missing in Obsewrvation 41951
# of data Features has missing Value 538


###### removing irrelavent columns

In [31]:
data_analysis.columns

Index(['dimension_Time', 'dimension_0', 'dimension_1', 'dimension_2',
       'dimension_3', 'dimension_4', 'dimension_5', 'dimension_6',
       'dimension_7', 'dimension_8',
       ...
       'dimension_581', 'dimension_582', 'dimension_583', 'dimension_584',
       'dimension_585', 'dimension_586', 'dimension_587', 'dimension_588',
       'dimension_589', 'Target'],
      dtype='object', length=592)

In [32]:
data_analysis.drop(columns=["dimension_Time"],inplace=True) # removing Time attribute

In [ ]:
# removing feature with same value for entire observation
data_stat = pd.DataFrame(data = data_analysis.describe(include="all").T.reset_index())
feature_name_with_same_value = list(data_stat[data_stat["min"] == data_stat["max"]]["index"])
#print(data_stat[data_stat["index"].isin(feature_name_with_same_value)])
data_analysis.drop(columns=feature_name_with_same_value,inplace=True)


In [58]:
# # Check for feature which has more than 75% of missing value
# missing_data = pd.DataFrame(data_analysis.isna().sum().reset_index(),columns=["Feature_name","Missing_count"])
# missing_data["per_missing"] = (missing_data["Missing_count"] / data_analysis.shape[0])

In [63]:
missing_data = pd.DataFrame(data_analysis.isna().sum().reset_index(),columns=['a', 'b', 'c'])
missing_data

a   b   c
0   NaN NaN NaN
1   NaN NaN NaN
2   NaN NaN NaN
3   NaN NaN NaN
4   NaN NaN NaN
5   NaN NaN NaN
6   NaN NaN NaN
7   NaN NaN NaN
8   NaN NaN NaN
9   NaN NaN NaN
10  NaN NaN NaN
11  NaN NaN NaN
12  NaN NaN NaN
13  NaN NaN NaN
14  NaN NaN NaN
15  NaN NaN NaN
16  NaN NaN NaN
17  NaN NaN NaN
18  NaN NaN NaN
19  NaN NaN NaN
20  NaN NaN NaN
21  NaN NaN NaN
22  NaN NaN NaN
23  NaN NaN NaN
24  NaN NaN NaN
25  NaN NaN NaN
26  NaN NaN NaN
27  NaN NaN NaN
28  NaN NaN NaN
29  NaN NaN NaN
30  NaN NaN NaN
31  NaN NaN NaN
32  NaN NaN NaN
33  NaN NaN NaN
34  NaN NaN NaN
35  NaN NaN NaN
36  NaN NaN NaN
37  NaN NaN NaN
38  NaN NaN NaN
39  NaN NaN NaN
40  NaN NaN NaN
41  NaN NaN NaN
42  NaN NaN NaN
43  NaN NaN NaN
44  NaN NaN NaN
45  NaN NaN NaN
46  NaN NaN NaN
47  NaN NaN NaN
48  NaN NaN NaN
49  NaN NaN NaN
50  NaN NaN NaN
51  NaN NaN NaN
52  NaN NaN NaN
53  NaN NaN NaN
54  NaN NaN NaN
55  NaN NaN NaN
56  NaN NaN NaN
57  NaN NaN NaN
58  NaN NaN NaN
59  NaN NaN NaN
60  NaN NaN NaN
61  NaN NaN NaN
62  NaN NaN NaN
63  NaN NaN NaN
64  NaN NaN NaN
65  NaN NaN NaN
66  NaN NaN NaN
67  NaN NaN NaN
68  NaN NaN NaN
69  NaN NaN NaN
70  NaN NaN NaN
71  NaN NaN NaN
72  NaN NaN NaN
73  NaN NaN NaN
74  NaN NaN NaN
75  NaN NaN NaN
76  NaN NaN NaN
77  NaN NaN NaN
78  NaN NaN NaN
79  NaN NaN NaN
80  NaN NaN NaN
81  NaN NaN NaN
82  NaN NaN NaN
83  NaN NaN NaN
84  NaN NaN NaN
85  NaN NaN NaN
86  NaN NaN NaN
87  NaN NaN NaN
88  NaN NaN NaN
89  NaN NaN NaN
90  NaN NaN NaN
91  NaN NaN NaN
92  NaN NaN NaN
93  NaN NaN NaN
94  NaN NaN NaN
95  NaN NaN NaN
96  NaN NaN NaN
97  NaN NaN NaN
98  NaN NaN NaN
99  NaN NaN NaN
100 NaN NaN NaN
101 NaN NaN NaN
102 NaN NaN NaN
103 NaN NaN NaN
104 NaN NaN NaN
105 NaN NaN NaN
106 NaN NaN NaN
107 NaN NaN NaN
108 NaN NaN NaN
109 NaN NaN NaN
110 NaN NaN NaN
111 NaN NaN NaN
112 NaN NaN NaN
113 NaN NaN NaN
114 NaN NaN NaN
115 NaN NaN NaN
116 NaN NaN NaN
117 NaN NaN NaN
118 NaN NaN NaN
119 NaN NaN NaN
120 NaN NaN NaN
121 NaN NaN NaN
122 NaN NaN NaN
123 NaN NaN NaN
124 NaN NaN NaN
125 NaN NaN NaN
126 NaN NaN NaN
127 NaN NaN NaN
128 NaN NaN NaN
129 NaN NaN NaN
130 NaN NaN NaN
131 NaN NaN NaN
132 NaN NaN NaN
133 NaN NaN NaN
134 NaN NaN NaN
135 NaN NaN NaN
136 NaN NaN NaN
137 NaN NaN NaN
138 NaN NaN NaN
139 NaN NaN NaN
140 NaN NaN NaN
141 NaN NaN NaN
142 NaN NaN NaN
143 NaN NaN NaN
144 NaN NaN NaN
145 NaN NaN NaN
146 NaN NaN NaN
147 NaN NaN NaN
148 NaN NaN NaN
149 NaN NaN NaN
150 NaN NaN NaN
151 NaN NaN NaN
152 NaN NaN NaN
153 NaN NaN NaN
154 NaN NaN NaN
155 NaN NaN NaN
156 NaN NaN NaN
157 NaN NaN NaN
158 NaN NaN NaN
159 NaN NaN NaN
160 NaN NaN NaN
161 NaN NaN NaN
162 NaN NaN NaN
163 NaN NaN NaN
164 NaN NaN NaN
165 NaN NaN NaN
166 NaN NaN NaN
167 NaN NaN NaN
168 NaN NaN NaN
169 NaN NaN NaN
170 NaN NaN NaN
171 NaN NaN NaN
172 NaN NaN NaN
173 NaN NaN NaN
174 NaN NaN NaN
175 NaN NaN NaN
176 NaN NaN NaN
177 NaN NaN NaN
178 NaN NaN NaN
179 NaN NaN NaN
180 NaN NaN NaN
181 NaN NaN NaN
182 NaN NaN NaN
183 NaN NaN NaN
184 NaN NaN NaN
185 NaN NaN NaN
186 NaN NaN NaN
187 NaN NaN NaN
188 NaN NaN NaN
189 NaN NaN NaN
190 NaN NaN NaN
191 NaN NaN NaN
192 NaN NaN NaN
193 NaN NaN NaN
194 NaN NaN NaN
195 NaN NaN NaN
196 NaN NaN NaN
197 NaN NaN NaN
198 NaN NaN NaN
199 NaN NaN NaN
200 NaN NaN NaN
201 NaN NaN NaN
202 NaN NaN NaN
203 NaN NaN NaN
204 NaN NaN NaN
205 NaN NaN NaN
206 NaN NaN NaN
207 NaN NaN NaN
208 NaN NaN NaN
209 NaN NaN NaN
210 NaN NaN NaN
211 NaN NaN NaN
212 NaN NaN NaN
213 NaN NaN NaN
214 NaN NaN NaN
215 NaN NaN NaN
216 NaN NaN NaN
217 NaN NaN NaN
218 NaN NaN NaN
219 NaN NaN NaN
220 NaN NaN NaN
221 NaN NaN NaN
222 NaN NaN NaN
223 NaN NaN NaN
224 NaN NaN NaN
225 NaN NaN NaN
226 NaN NaN NaN
227 NaN NaN NaN
228 NaN NaN NaN
229 NaN NaN NaN
230 NaN NaN NaN
231 NaN NaN NaN
232 NaN NaN NaN
233 NaN NaN NaN
234 NaN NaN NaN
235 NaN NaN NaN
236 NaN NaN NaN
237 NaN NaN NaN
238 NaN NaN NaN
239 NaN NaN NaN
240 NaN NaN NaN
241 NaN NaN NaN
242 NaN NaN NaN
243 NaN NaN NaN
244 NaN NaN NaN
245 NaN NaN NaN
246 NaN NaN NaN
247 NaN NaN NaN
248 NaN NaN NaN
249 Na

In [59]:
missing_data

Feature_name  Missing_count  per_missing
0             NaN            NaN          NaN
1             NaN            NaN          NaN
2             NaN            NaN          NaN
3             NaN            NaN          NaN
4             NaN            NaN          NaN
5             NaN            NaN          NaN
6             NaN            NaN          NaN
7             NaN            NaN          NaN
8             NaN            NaN          NaN
9             NaN            NaN          NaN
10            NaN            NaN          NaN
11            NaN            NaN          NaN
12            NaN            NaN          NaN
13            NaN            NaN          NaN
14            NaN            NaN          NaN
15            NaN            NaN          NaN
16            NaN            NaN          NaN
17            NaN            NaN          NaN
18            NaN            NaN          NaN
19            NaN            NaN          NaN
20            NaN            NaN          NaN
21            NaN            NaN          NaN
22            NaN            NaN          NaN
23            NaN            NaN          NaN
24            NaN            NaN          NaN
25            NaN            NaN          NaN
26            NaN            NaN          NaN
27            NaN            NaN          NaN
28            NaN            NaN          NaN
29            NaN            NaN          NaN
30            NaN            NaN          NaN
31            NaN            NaN          NaN
32            NaN            NaN          NaN
33            NaN            NaN          NaN
34            NaN            NaN          NaN
35            NaN            NaN          NaN
36            NaN            NaN          NaN
37            NaN            NaN          NaN
38            NaN            NaN          NaN
39            NaN            NaN          NaN
40            NaN            NaN          NaN
41            NaN            NaN          NaN
42            NaN            NaN          NaN
43            NaN            NaN          NaN
44            NaN            NaN          NaN
45            NaN            NaN          NaN
46            NaN            NaN          NaN
47            NaN            NaN          NaN
48            NaN            NaN          NaN
49            NaN            NaN          NaN
50            NaN            NaN          NaN
51            NaN            NaN          NaN
52            NaN            NaN          NaN
53            NaN            NaN          NaN
54            NaN            NaN          NaN
55            NaN            NaN          NaN
56            NaN            NaN          NaN
57            NaN            NaN          NaN
58            NaN            NaN          NaN
59            NaN            NaN          NaN
60            NaN            NaN          NaN
61            NaN            NaN          NaN
62            NaN            NaN          NaN
63            NaN            NaN          NaN
64            NaN            NaN          NaN
65            NaN            NaN          NaN
66            NaN            NaN          NaN
67            NaN            NaN          NaN
68            NaN            NaN          NaN
69            NaN            NaN          NaN
70            NaN            NaN          NaN
71            NaN            NaN          NaN
72            NaN            NaN          NaN
73            NaN            NaN          NaN
74            NaN            NaN          NaN
75            NaN            NaN          NaN
76            NaN            NaN          NaN
77            NaN            NaN          NaN
78            NaN            NaN          NaN
79            NaN            NaN          NaN
80            NaN            NaN          NaN
81            NaN            NaN          NaN
82            NaN            NaN          NaN
83            NaN            NaN          NaN
84            NaN            NaN          NaN
85            NaN            NaN          NaN
86 

#### Validation Set preparation

In [ ]:
def check_distribution(traindata,testdata,columns_list):
    for column_name in columns_list:
        print("---------------",column_name,"------------")
        print(round(traindata[column_name].value_counts() / traindata[column_name].count(),2)*100)
        print(round(testdata[column_name].value_counts() / testdata[column_name].count(),2)*100)

In [ ]:
X_train_test_validate = data_impute.iloc[:,:-1]
y_train_test_validate = data_impute.iloc[:,-1:]

X_train_test, X_validate,y_train_test,y_validate = train_test_split(X_train_test_validate, y_train_test_validate, test_size=0.15, random_state=123)
check_distribution(y_train_test,y_validate,['Target'])


--------------- Target ------------
1    93.0
0     7.0
Name: Target, dtype: float64
1    94.0
0     6.0
Name: Target, dtype: float64
